circuit/collections/src/merkle_tree/mod.rs

```diff
@@ -67,16 +67,196 @@ impl<E: Environment, const DEPTH: u8> Eject for MerklePath<E, DEPTH> {
     }
 }
 
+#[derive(Clone)]
+pub struct MerkleTree<E: Environment, LH: LeafHash<E, Hash = PH::Hash>, PH: PathHash<E, Hash = Field<E>>, const DEPTH: u8> {
+    /// The leaf hasher for the Merkle tree.
+    leaf_hasher: LH,
+    /// The path hasher for the Merkle tree.
+    path_hasher: PH,
+    /// The computed root of the full Merkle tree.
+    root: PH::Hash,
+    /// The internal hashes, from root to hashed leaves, of the full Merkle tree.
+    tree: Vec<PH::Hash>,
+    /// The canonical empty hash.
+    empty_hash: Field<E>,
+    /// The number of hashed leaves in the tree.
+    number_of_leaves: usize,
+}
+
+impl<E: Environment, LH: LeafHash<E, Hash = PH::Hash>, PH: PathHash<E, Hash = Field<E>>, const DEPTH: u8> MerkleTree<E, LH, PH, DEPTH> {
+
+    #[inline]
+    /// Initializes a new Merkle tree with the given leaves.
+    pub fn new(leaf_hasher: LH, path_hasher: PH, leaves: &[LH::Leaf]) -> Result<Self> {
```

Comment:
For my own understanding, are there any differences between this implementation and the console one?